language/vm/src/views.rs

```diff
@@ -531,63 +532,120 @@ impl<'a, T: ModuleAccess> SignatureTokenView<'a, T> {
-                // - Otherwise if any type argument is `Kind::Resource`, returns `Kind::Resource`
-                // - Otherwise returns `Kind::Unrestricted`
-                type_arguments.iter().map(
-                    |token| Self::new(self.module, token).kind()
-                ).fold(Kind::Unrestricted, |acc_kind, next_kind| {
-                    match (acc_kind, next_kind) {
-                        (Kind::All, _) | (_, Kind::All) => Kind::All,
-                        (Kind::Resource, _) | (_, Kind::Resource) => Kind::Resource,
-                        (Kind::Unrestricted, Kind::Unrestricted) => Kind::Unrestricted
-                    }
-                })
+                // At this point all type actuals have been verified.
+                // It is safe to conclude early the struct (instance) is a valid resource if marked
+                // so.
+                if sh.is_nominal_resource {
+                    return Ok(Kind::Resource);
+                }
+
+                // Derive the kind of the struct (instance).
+                //   - If any of the type actuals have kind `all`, then the struct have kind `all`.
+                //     - `all` means part of the type can be either `resource` or `unrestricted`.
+                //     - Therefore it is also impossible to determine the kind of the type as a
+                //       whole, and thus `all`.
+                //   - If none of the type actuals has kind `all`, then the struct is a resource if
+                //     and only if one of the type actuuals has kind `resource`.
+                let kind = kinds
+                    .iter()
+                    .fold(Kind::Unrestricted, |acc_kind, next_kind| {
+                        match (acc_kind, next_kind) {
+                            (Kind::All, _) | (_, Kind::All) => Kind::All,
+                            (Kind::Resource, _) | (_, Kind::Resource) => Kind::Resource,
+                            (Kind::Unrestricted, Kind::Unrestricted) => Kind::Unrestricted,
+                        }
+                    });
+
+                Ok(kind)
             }
-            SignatureToken::Reference(_)
-            | SignatureToken::MutableReference(_)
-            | SignatureToken::Bool
-            | SignatureToken::U64
-            | SignatureToken::String
-            | SignatureToken::ByteArray
-            | SignatureToken::Address => Kind::Unrestricted,
-            // TODO: To get the kind of a type parameter we need to look at the struct/function
-            // that contains it. Change the API or remodel accesses/views with a tiered system.
-            SignatureToken::TypeParameter(_) => panic!("cannot tell if a type parameter is a resource or not (feature not yet implemented)"),
         }
     }
 
-    pub fn contains_nominal_resource(&self) -> bool {
+    // TODO: refactor views so that we get the type formals from self.
+    pub fn contains_nominal_resource(&self, type_formals: &[Kind]) -> bool {
```

Comment:
I am confused between type_actuals and type_formals.  The deleted comment mentions type_actuals but the parameter is called type_formals.

Reply:
Type formals are `(type variable, kind)` declarations in the function signature:
```
foo<T1, T2: resource>()
   ^ type formals
```
Type actuals are types you use to instantiate an instance:
```
foo<u64, LibraCoin.T>
   ^ type actuals
```

Reply:
I will wait to see the English documentation for fn kind and fn contains_nominal_resource before I comment further on the suitability of type_formals vs. type_actuals as the name of the parameter.

Reply:
There's been another proposal suggesting (type_parameters & type_arguments) vs (type_formals & type_actuals). I personally don't care, as long as they are consistent everywhere. Anyway I'll add the comments.

Reply:
> I will wait to see the English documentation for fn kind and fn contains_nominal_resource before I comment further on the suitability of type_formals vs. type_actuals as the name of the parameter.

Ah that makes sense. I've added the comments and made it clear that the parameter `type_formals` is part of the context required to determine the kind of a type variable. It is set to either `function_signature.type_formals` or `struct_handle.type_formals`. 

I have to admit this is quite messy, but doing it the right way would require a pretty big refactor of views as a whole, which will divert us from our focus on generics. I'll likely do it in a short term.

Please ignore my previous comment. I seem to have misunderstood what you were saying.